### PR TITLE
Cura 11003 add searchbar for offlineprinter

### PR DESCRIPTION
**Pull request by saumyaj3**
# Description

Add Searchbar to the "Add An Offline Printer" list
search

### CODE REVIEW

**Review by wawanbreton — COMMENTED** [review]
Just some minor remarks and questions.

I also wanted to point out that Qt has a `QSortFilterProxyModel` class that is designed to handle this use case. You don't have to change the data in the model, but just apply the filter on the view. This is a bit faster and also makes it much easier to keep the selection and hide empty sections. I have a real-life example [here](https://github.com/wawanbreton/cvcomposer/blob/461067141185e28f7fed6d3d9c2bbfbac4198c28/gui/mainwidget.cpp#L51) if you are interested (in C++ though)

File: resources/qml/WelcomePages/AddLocalPrinterScrollView.qml [hunk 1]
```diff
@@ -15,7 +15,7 @@ import Cura 1.1 as Cura
 Item
 {
     id: base
-
+    property bool findingPrinters: false
```
[the remark this answers is not in the document]
Comment (saumyaj3) [reply resources/qml/WelcomePages/AddLocalPrinterScrollView.qml:18]: something like searchFilterApplied?

[the remark this answers is not in the document]
Comment (wawanbreton) [reply resources/qml/WelcomePages/AddLocalPrinterScrollView.qml:18]: Hmm that also makes me think that there is something dynamic related, like "the search filter has now been applied". What about "hasSearchFilter" ? Sorry I'm being nitpicking...

File: resources/qml/WelcomePages/AddLocalPrinterScrollView.qml [hunk 2]
```diff
@@ -43,25 +43,31 @@ Item
...
-        if (metadata)
-        {
-            return metadata[key];
-        }
-        return undefined;
+        id: printerSearchTimer
+        onTriggered: filter.editingFinished()
+        interval: 500
+        running: false
+        repeat: false
```
[the remark this answers is not in the document]
Comment (saumyaj3) [reply resources/qml/WelcomePages/AddLocalPrinterScrollView.qml:70]: changed it to 50ms

File: resources/qml/WelcomePages/AddLocalPrinterScrollView.qml [hunk 3]
```diff
@@ -73,6 +79,93 @@ Item
         base.currentSections = base.currentSections;
     }
 
+    Cura.TextField
+    {
+        id: filter
+        width: Math.floor(parent.width * 0.48)
```
[the remark this answers is not in the document]
Comment (saumyaj3) [reply resources/qml/WelcomePages/AddLocalPrinterScrollView.qml]: yes done in[2ed0377](https://github.com/Ultimaker/Cura/pull/18804/commits/2ed0377a052586fafa5557bbaa79c2e01313c33c)

File: resources/qml/WelcomePages/AddLocalPrinterScrollView.qml (hunk 1 above)
[the remark this answers is not in the document]
Comment (saumyaj3) [reply resources/qml/WelcomePages/AddLocalPrinterScrollView.qml]: That's alright. it should be understandable to all. [f35865e](https://github.com/Ultimaker/Cura/pull/18804/commits/f35865e3c48fcd57b9768c176bf41ec4ad79d65b)

File: resources/qml/WelcomePages/AddLocalPrinterScrollView.qml (hunk 2 above)
[the remark this answers is not in the document]
Comment (wawanbreton) [reply resources/qml/WelcomePages/AddLocalPrinterScrollView.qml:70]: That should do it, but why use a Timer at all ? This is all done on QML side so there should be no threading issue. :thinking:

[the remark this answers is not in the document]
Comment (wawanbreton) [reply resources/qml/WelcomePages/AddLocalPrinterScrollView.qml:70]: Have you tried not using a timer ? Does it generate some issue ?

[the remark this answers is not in the document]
Comment (saumyaj3) [reply resources/qml/WelcomePages/AddLocalPrinterScrollView.qml]: Done

[the remark this answers is not in the document]
Comment (wawanbreton) [reply resources/qml/WelcomePages/AddLocalPrinterScrollView.qml]: Awesome :smiley: Less code, less bugs !

**Review by wawanbreton — APPROVED** [review]

File: resources/qml/WelcomePages/AddLocalPrinterScrollView.qml [hunk 4]
```diff
@@ -78,98 +81,204 @@ Item
...
+                    id: filter
+                    width:parent.width
+                    implicitHeight: parent.height
+                    background: Rectangle {
+                        id: background
+                        color: UM.Theme.getColor("main_background")
+                        radius: 2
+                        border.color:  UM.Theme.getColor("primary_button")
+                    }
+                    height: UM.Theme.getSize("small_button_icon").height*2
```
Comment (nallath) [inline resources/qml/WelcomePages/AddLocalPrinterScrollView.qml:104]:
```suggestion
                    height: UM.Theme.getSize("small_button_icon").height * 2
```

File: resources/qml/WelcomePages/AddLocalPrinterScrollView.qml [hunk 5]
```diff
@@ -78,98 +81,204 @@ Item
...
+                height: filter.height
+                Cura.TextField
+                {
+                    id: filter
+                    width:parent.width
+                    implicitHeight: parent.height
+                    background: Rectangle {
+                        id: background
+                        color: UM.Theme.getColor("main_background")
+                        radius: 2
```
Comment (nallath) [inline resources/qml/WelcomePages/AddLocalPrinterScrollView.qml:101]: Shouldn't this also be from a theme?

File: resources/qml/WelcomePages/AddLocalPrinterScrollView.qml [hunk 6]
```diff
@@ -78,98 +81,204 @@ Item
...
+                        border.color:  UM.Theme.getColor("primary_button")
```
Comment (nallath) [inline resources/qml/WelcomePages/AddLocalPrinterScrollView.qml:102]:
```suggestion
                        border.color: UM.Theme.getColor("primary_button")
```

File: resources/qml/WelcomePages/AddLocalPrinterScrollView.qml [hunk 7]
```diff
@@ -78,98 +81,204 @@ Item
...
             width: Math.floor(parent.width * 0.48)
             height: parent.height
+            Item
+            {
+                width: root.width
+                height: filter.height
+                Cura.TextField
+                {
+                    id: filter
+                    width:parent.width
```
Comment (nallath) [inline resources/qml/WelcomePages/AddLocalPrinterScrollView.qml:96]:
```suggestion
                    width: parent.width
```

File: resources/qml/WelcomePages/AddLocalPrinterScrollView.qml [hunk 8]
```diff
@@ -78,98 +81,204 @@ Item
...
-            section.delegate: Button
+            //Selecting a local printer to add from this list.
+            ListView
             {
-                id: button
-                width: machineList.width
-                height: UM.Theme.getSize("action_button").height
-                text: section
+                id: machineList
+                width:root.width
```
Comment (nallath) [inline resources/qml/WelcomePages/AddLocalPrinterScrollView.qml:194]:
```suggestion
                width: root.width
```

Comment (nallath) [inline resources/qml/WelcomePages/AddLocalPrinterScrollView.qml:194]:
```suggestion
                width: root.width
```

**Review by nallath — CHANGES_REQUESTED** [review]

File: resources/qml/WelcomePages/AddLocalPrinterScrollView.qml (hunk 4 above)
Comment (saumyaj3) [reply resources/qml/WelcomePages/AddLocalPrinterScrollView.qml]: a3911bdbfa47b0825d9b4980584fa9cf37ed52a1

File: resources/qml/WelcomePages/AddLocalPrinterScrollView.qml (hunk 5 above)
Comment (saumyaj3) [reply resources/qml/WelcomePages/AddLocalPrinterScrollView.qml]: a3911bdbfa47b0825d9b4980584fa9cf37ed52a1

File: resources/qml/WelcomePages/AddLocalPrinterScrollView.qml (hunk 6 above)
Comment (saumyaj3) [reply resources/qml/WelcomePages/AddLocalPrinterScrollView.qml]: a3911bdbfa47b0825d9b4980584fa9cf37ed52a1

**Comment by johnsmith491** [comment]
has anyone tried this method? and how it is performing?

**Comment by saumyaj3** [comment]
solves https://github.com/Ultimaker/Cura/issues/16716. This feature will be available in the next version of cura.

**Pull request merged**